rmw_zenoh_cpp/src/detail/zenoh_router_check.cpp

```diff
@@ -0,0 +1,88 @@
+// Copyright 2023 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "zenoh_router_check.hpp"
+
+#include <rcutils/env.h>
+#include <rcutils/logging_macros.h>
+
+#include <string>
+#include <sstream>
+#include <iomanip>
+
+namespace
+{
+
+// Convert a Zenoh Id to a string
+// Zenoh IDs are LSB-first 128bit unsigned and non-zero integers in hexadecimal lowercase.
+// @param pid Zenoh Id to convert
+std::string ZidToStr(z_id_t pid)
+{
+  std::stringstream ss;
+  int len = 0;
+  for (int i = 0; i < 16; i++) {
+    if (pid.id[i]) {
+      len = i + 1;
+    }
+  }
+  if (!len) {
+    ss << "";
+  } else {
+    for (int i = len - 1; i >= 0; --i) {
+      ss << std::hex << std::setfill('0') << std::setw(2) << static_cast<int>(pid.id[i]);
+    }
+  }
+  return ss.str();
+}
+
+}  // namespace
+
+rmw_ret_t zenoh_router_check(z_session_t session)
+{
+  // Initialize context for callback
+  void * context = malloc(sizeof(int));
+  *(static_cast<int *>(context)) = 0;
+
+  // Define callback
+  auto callback = [](const struct z_id_t * id, void * ctx) {
+      const std::string id_str = ZidToStr(*id);
+      RCUTILS_LOG_INFO_NAMED(
+        "ZenouRouterCheck",
+        "A Zenoh router connected to the session with id '%s'", id_str.c_str());
+      // Note: Callback is guaranteed to never be called
+      // concurrently according to z_info_routers_zid docstring
+      (*(static_cast<int *>(ctx)))++;
+    };
+
+  z_owned_closure_zid_t router_callback = z_closure(callback, nullptr /* drop */, context);
+  z_info_routers_zid(session, z_move(router_callback));
+
+  rmw_ret_t ret;
+  if (*(static_cast<int *>(context)) == 0) {
```

Comment:
I'm trying to understand how the `z_info_routers_zid()` API works.
```
/**
 * Fetches the Zenoh IDs of all connected routers.
 *
 * `callback` will be called once for each ID, is guaranteed to never be called concurrently,
 * and is guaranteed to be dropped before this function exits.
 *
 * Retuns 0 on success, negative values on failure.
 */
ZENOHC_API
int8_t z_info_routers_zid(struct z_session_t session,
                          struct z_owned_closure_zid_t *callback);
 ```
 
 If it returns an `int8_t`, is it a blocking function or will the function execute callbacks asynchronously? If its the former, we could just check if the return value is 0 and if so return `RMW_RET_OK`? Unless we plan to parse the zids we receive in the callbacks to check if the zid matches that of the zenoh router we run (and not any other zenoh router)?

Reply:
> is it a blocking function or will the function execute callbacks asynchronously? 

It is a blocking function.

> we could just check if the return value is 0 and if so return RMW_RET_OK?

It "returns 0 in success" but that doesn't convey whether there are or not available routers. It would return 0 even if the zenoh router is missing. This only states whether the execution was correct. (Just realized that we should return error in case there is a failure, adding that...)

Reply:
> Unless we plan to parse the zids we receive in the callbacks to check if the zid matches that of the zenoh router we run (and not any other zenoh router)?

For the moment we need the callback to check that at least a router is connected to the session. I added the parsing of the id mainly for logging purposes. We could at some point verify the id to correctly assert that we are connected to the zenoh router we want. However it is a bit early for this I think, for the moment we don't have environments with more than one zenoh router.